src/ogr/geometry.jl

```diff
@@ -1215,42 +1223,70 @@ function getgeom(geom::AbstractGeometry, i::Integer)::IGeometry
-        ArchGDAL.IGeometry{ArchGDAL.wkbLineString},
-        ArchGDAL.Geometry{ArchGDAL.wkbLineString},
+        IGeometry{wkbLineString},
+        Geometry{wkbLineString},
     }, # TODO All curves
     i::Integer,
 )::IGeometry
-    if geom.ptr == C_NULL
-        return IGeometry()
-    end
+    geom.ptr == C_NULL && return IGeometry{wkbUnknown}()
     return createpoint(getpoint(geom, i)[1:getcoorddim(geom)])
 end
-
 function unsafe_getgeom(geom::AbstractGeometry, i::Integer)::Geometry
     # NOTE(yeesian): GDAL.ogr_g_getgeometryref(geom, i) returns an handle to a
     # geometry within the container. The returned geometry remains owned by the
     # container, and should not be modified. The handle is only valid until the
     # next change to the geometry container. Use OGR_G_Clone() to make a copy.
-    if geom.ptr == C_NULL
-        return Geometry()
-    end
+    geom.ptr == C_NULL && return Geometry{wkbUnknown}()
     result = GDAL.ogr_g_getgeometryref(geom.ptr, i)
-    if result == C_NULL
-        return Geometry()
-    else
-        return Geometry(GDAL.ogr_g_clone(result))
+    result == C_NULL && return Geometry{wkbUnknown}()
+    return Geometry(GDAL.ogr_g_clone(result))
+end
+# Specialised methods for type stability
+# Where we know the child geometry we use the enum in the
+# type explicitly
+for (parent, child) in (
+    (wkbPolygon, wkbLineString),
+    (wkbPolygonM, wkbLineStringM),
+    (wkbPolygonZM, wkbLineStringZM),
+    (wkbPolygon25D, wkbLineString25D),
+    (wkbLineString, wkbPoint),
+    (wkbLineStringM, wkbPointM),
+    (wkbLineStringZM, wkbPointZM),
+    (wkbLineString25D, wkbPoint25D),
+    (wkbMultiLineString, wkbLineString),
+    (wkbMultiLineStringM, wkbLineStringM),
+    (wkbMultiLineStringZM, wkbLineStringZM),
+    (wkbMultiLineString25D, wkbLineString25D),
+    (wkbMultiPoint, wkbPoint),
+    (wkbMultiPointM, wkbPointM),
+    (wkbMultiPointZM, wkbPointZM),
+    (wkbMultiPoint25D, wkbPoint25D),
+    (wkbMultiPolygon, wkbPolygon),
+    (wkbMultiPolygonM, wkbPolygonM),
+    (wkbMultiPolygonZM, wkbPolygonZM),
+    (wkbMultiPolygon25D, wkbPolygon25D),
+    (wkbCurvePolygon, wkbCircularString),
+    (wkbCurvePolygonM, wkbCircularStringM),
+    (wkbCurvePolygonZM, wkbCircularStringZM),
+)
+    @eval function getgeom(geom::AbstractGeometry{$parent}, i::Integer)::Union{IGeometry{wkbUnknown},IGeometry{$child}}
+        geom.ptr == C_NULL && return IGeometry{wkbUnknown}()
+        result = GDAL.ogr_g_getgeometryref(geom.ptr, i)
+        result == C_NULL && return IGeometry{wkbUnknown}()
+        return IGeometry{$child}(GDAL.ogr_g_clone(result))
+    end
+    @eval function unsafe_getgeom(geom::AbstractGeometry{$parent}, i::Integer)::Union{Geometry{wkbUnknown},Geometry{$child}}
+        geom.ptr == C_NULL && return Geometry{wkbUnknown}()
+        result = GDAL.ogr_g_getgeometryref(geom.ptr, i)
+        result == C_NULL && return Geometry{wkbUnknown}()
+        return Geometry{$child}(GDAL.ogr_g_clone(result))
     end
```

Comment:
The package doesn't have test coverage for this yet.

Reply:
I've added tests for these.